### DISCUSSION TITLE
The pain of migrating to nx (from multiple CRA apps)

### DISCUSSION THREAD

**Discussion by amivag** [nrwl/nx#15351] [General]
Hello,
I would like to share my recent experience with Nx, and the major problems I faced with migrating my React apps, in the hopes that it will help improve things in the future...

I have several React apps (built using CRA), which use similar code, and I wanted to bring them all under a monorepo for easy code reuse. I have no prior experience with monorepos, I've been studying youtube tutorials, and it sounded exactly like what I needed.

nx was the most recommended library so I went with that.

It has been an incredibly painful experience so far. These are the most important reasons:

1) I was getting errors even at the most basic actions. The CRA to nx migration command, was failing on me. It would not complete the process. I had taken care to update the app to latest React + components before migration. The issue was resolved by an nx update, shortly after I opened an issue on github, apparently it was an issue with windows and path handling (after a lot of hair pulling on my part, I was not expecting a failure that early, on a library already on version 15.x).

2) As I said, I was getting errors often. In fact, the chances are that I would get an error was higher, compared to the command simply working as intended. I was getting another error by attempting to remove a library, and the nx warning me that all my apps would have to be removed as well (!). The issue was resolved after I opened a github issue with the details (kudos for that... but again... this is v15.x)

3) The are different terms referring to similar things, all over the place. The tutorials talked about apps and libs folder (which sounded relatively straightforward). But then, the docs refer to integrated or package repos, and talk about different folder structures and terminologies. And then there was a standalone app mode. I was confused. Still am. 

4) I migrated my app to nx, in the hopes that it would end up in an apps folder... only to find it in the root folder as a single app. What is the point of that?

5) The nx commands, often asked me several questions before completing an action, which were not covered in the docs or the tutorials. Like e.g. what builder to use for a new library (vite, webpack, rollup etc). What do I choose? Do I even need to choose? Help.

6) In the docs there was often a mention for an important file called workspace.json. I do not have such a file. Is my nx repo corrupted? Help.

7) Why do I have to use a "move" command to rename a lib?

8) I use @nrwl/react something to create a React lib, but I need to use a @nrwl/workspace or a @nrwl/react to rename or move a lib? I am not sure. Commands are terribly confusing (didn't have much better luck with the VSCode nx console, though I am working on this).

9) I don't have a complete understanding of a good, recommended or optimal workflow for structuring my react apps and libs. And it's very hard to find clear documentation on this. (are libs standalone components? do they host multiple react components? what is a recommended directory structure? etc)

I will end this here. I so far have only managed to half-migrate my apps into separate nx repositories... hoping one day I will be able to have them all under a single nx project, working as intended... working on it, slowly, patiently... 

No need to reply to my questions, I am just giving examples of problems I faced. I am hoping things get much improved in the future.

Thanks for your time!

**Comment by amartincolby** [comment]
The different directory structures were a point of confusion about which I just posted. Did you every find clarity on that and your other questions? This is high-value information.
